src/renderer/atlas/CustomGlyphs.cpp

```diff
@@ -0,0 +1,1265 @@
+
+        begX = roundf(begX - lineOffsetX) + lineOffsetX;
+        begY = roundf(begY - lineOffsetY) + lineOffsetY;
+        endX = roundf(endX + lineOffsetX) - lineOffsetX;
+        endY = roundf(endY + lineOffsetY) - lineOffsetY;
+
+        switch (shape)
+        {
+        case Shape_LightLine:
+        case Shape_HeavyLine:
+        {
+            const D2D1_POINT_2F beg{ begX, begY };
+            const D2D1_POINT_2F end{ endX, endY };
+            renderTarget->DrawLine(beg, end, brush, lineWidth, nullptr);
+            break;
+        }
+        case Shape_EmptyRect:
+        case Shape_RoundRect:
+        {
+            const D2D1_ROUNDED_RECT r{ { begX, begY, endX, endY }, lightLineWidth * 2, lightLineWidth * 2 };
+            if (shape == Shape_EmptyRect)
+            {
+                renderTarget->DrawRectangle(&r.rect, brush, lineWidth, nullptr);
+            }
+            else
+            {
+                renderTarget->DrawRoundedRectangle(&r, brush, lineWidth, nullptr);
+            }
+            break;
+        }
+        case Shape_FilledEllipsis:
+        case Shape_EmptyEllipsis:
+        {
+            const D2D1_ELLIPSE e{ { begX, begY }, endX, endY };
+            if (shape == Shape_FilledEllipsis)
+            {
+                renderTarget->FillEllipse(&e, brush);
+            }
+            else
+            {
+                renderTarget->DrawEllipse(&e, brush, lineWidth, nullptr);
+            }
+            break;
+        }
+        case Shape_ClosedFilledPath:
+        case Shape_OpenLinePath:
+            if (instruction.begX)
+            {
+                geometryPoints[geometryPointsCount++] = { begX, begY };
+            }
+            if (instruction.endX)
+            {
+                geometryPoints[geometryPointsCount++] = { endX, endY };
+            }
+            break;
+        default:
+        {
+            wil::com_ptr<ID2D1BitmapBrush> bitmapBrush;
+            ID2D1Brush* b = brush;
+
+            if (shape >= Shape_Filled075 && shape <= Shape_Filled025)
+            {
+                static constexpr u32 _ = 0;
+                static constexpr u32 w = 0xffffffff;
+                static constexpr u32 size = 4;
+                // clang-format off
+                static constexpr u32 shades[3][size * size] = {
+                    {
+                        w, w, w, _,
+                        w, w, _, w,
+                        w, _, w, w,
+                        _, w, w, w,
+                    },
+                    {
+                        w, _, w, _,
+                        _, w, _, w,
+                        w, _, w, _,
+                        _, w, _, w,
+                    },
+                    {
+                        _, _, _, w,
+                        _, _, w, _,
+                        _, w, _, _,
+                        w, _, _, _,
+                    },
+                };
```

Comment:
Okay, intially I thought _we are_ drawing lines instead of dots. 

I think we should scale the shades texture (relative to font-size `em`) so that `_` (gaps) and `w` (whites) are noticably large. Right now, the glyphs look like they're made up of highly densed slanted lines.

The texture pattern looks fine I guess, just need to scale it up.

**font glyphs**
![image](https://github.com/microsoft/terminal/assets/55626797/8f5498c3-88a8-45d8-9e0d-1b876f15b591)


**builtin glyphs**
![image](https://github.com/microsoft/terminal/assets/55626797/d0584d65-413c-4eb0-86ab-d8e62e6d8775)

Reply:
Initially I considered scaling the lines, but then I realized that making them as small as possible would have the benefit of being just... kinda pretty on current HiDPI displays. Today's 150-200 DPI display have just the right resolution where you can still barely see the lines but it also blurs with the background. I've tested this on a 200 DPI display and it looked kinda pretty.

That said, I want to scale the lines up as well, because otherwise it won't work on 250+ DPI displays in the future. :)

The reason I'm using slanted lines is because dot patterns like you mentioned can produce [moiré patterns](https://en.wikipedia.org/wiki/Moir%C3%A9_pattern) between neighboring cells. Like this:
![image](https://github.com/microsoft/terminal/assets/2256941/d7a01ad6-a11b-4259-ba5a-2d8c6eb9e681)

It can be solved by making the dot height proportional to the cell height to ensure that the last row of dots does not neighbor the surrounding cells. Your example font does this quite well. But slanted lines have a worst case moiré pattern that is way less obvious in general, so I think they're superior:
![image](https://github.com/microsoft/terminal/assets/2256941/72d301b0-f9fb-45ab-ac9e-e13cdd4d90f4)

The resulting image looks somewhat different, but I think this is fine (the red is less vibrant with the slanted lines). Terminal coloring is not really an exact since anyways.

Reply:
Oh hold on... This is also kinda pretty:
![image](https://github.com/microsoft/terminal/assets/2256941/401f85ad-4f6d-47c5-b9ff-dd473392bf60)

Inspired by your font:
```cpp
{
    w, _, _, _,
    w, _, _, _,
    _, _, w, _,
    _, _, w, _,
},
{
    w, _, w, _,
    _, w, _, w,
    w, _, w, _,
    _, w, _, w,
},
{
    _, w, w, w,
    _, w, w, w,
    w, w, _, w,
    w, w, _, w,
},
```

Reply:
> The reason I'm using slanted lines is because dot patterns like you mentioned can produce moiré patterns between neighboring cells.

This is an interesting point.

I think the current approach works just fine. I'm little concerned about user acceptance, like people who are used to seeing these Shade glyphs from their favourite font might find them _visually different_. However, I agree we can keep this for future PR 🙂